Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,248 @@
+
+        
+
+    def approximate_gradient(self, x, function_num, out=None):
+        """ Calculates the gradient of the selected function, indexed by `function_number` in {0,...,len(functions)-1}, at the point :math:`x` and at the stored snapshot :math:`\tilde{x}` before returning the stochastic gradient
+            .. math ::
+                \nabla f_i(x_k) - \nabla f_i(\tilde{x}) + \nabla \sum_{i=0}^{n-1}f_i(\tilde{x})
+        
+        Parameters
+        ----------
+        x : DataContainer ( e.g. ImageData)
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+        Returns
+        --------
+        DataContainer (e.g. ImageData)
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+    
+
+        self._svrg_iter_number += 1
+
+        self.stoch_grad_at_iterate = self.functions[function_num].gradient(x)
+
+        if self.store_gradients is True:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self._list_stored_gradients[function_num], -1.)
+        else:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self.functions[function_num].gradient(self.snapshot), -1.)
+
+        self._update_data_passes_indices([function_num])
+
+        if out is None:
+            out = self._stochastic_grad_difference.sapyb(
+                self.num_functions, self._full_gradient_at_snapshot, 1.)
+        else:
+            self._stochastic_grad_difference.sapyb(
+                self.num_functions, self._full_gradient_at_snapshot, 1., out=out)
+
+        return out
+
+    def _update_full_gradient_and_return(self, x, out=None):
+        """
+        Takes a "snapshot" at the point :math:`x`, saving both the point :math:` \tilde{x}=x` and its gradient :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}`. The function returns :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}` as the gradient calculation. If :code:`store_gradients==True`, the gradient of all the :math:`f_i`s is computed and stored at the "snapshot"..
+        
+        Parameters
+        ----------
+        x : DataContainer (e.g. ImageData)
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer (e.g. ImageData)
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+
+        self._svrg_iter_number += 1 
+
+        if self.store_gradients is True:
+            #Save the gradient of each individual f_i and the gradient of the full sum at the point x. 
+            self._list_stored_gradients = [
+                fi.gradient(x) for fi in self.functions]
```

Comment:
presumably better to re-use memory of previously stored gradients here`

Reply:
I think I updated this but it feels messier. Open to suggestions